### ISSUE TITLE
Reql_admin branch - [split_points->at(i) >= split_points->at(i-1)] 

### ISSUE THREAD

**Issue by neumino** [rethinkdb/rethinkdb#3020]
When trying to create something like 100 shards, I got this crash:

```
Version: rethinkdb 1.14.0-652-g9bd402-dirty (debug) (GCC 4.9.1)
error: Error in src/clustering/administration/tables/split_points.cc at line 72:
error: Guarantee failed: [split_points->at(i) >= split_points->at(i-1)] 
error: Backtrace:
error: Sun Sep  7 00:24:12 2014
       
       1: rethinkdb_backtrace(void**, int) at rethinkdb_backtrace.cc:101
       2: backtrace_t::backtrace_t() at backtrace.cc:203 (discriminator 1)
       3: lazy_backtrace_formatter_t::lazy_backtrace_formatter_t() at backtrace.cc:283
       4: format_backtrace(bool) at backtrace.cc:198
       5: report_fatal_error(char const*, int, char const*, ...) at errors.cc:83 (discriminator 1)
       6: ../build/debug/rethinkdb() [0x1ca5c0a] at 0x1ca5c0a ()
       7: calculate_split_points_with_distribution(uuid_u, real_reql_cluster_interface_t*, unsigned long, signal_t*, table_shard_scheme_t*, std::string*) at split_points.cc:151
       8: calculate_split_points_intelligently(uuid_u, real_reql_cluster_interface_t*, unsigned long, table_shard_scheme_t const&, signal_t*, table_shard_scheme_t*, std::string*) at split_points.cc:200
       9: real_reql_cluster_interface_t::table_reconfigure(counted_t<ql::db_t const>, name_string_t const&, table_generate_config_params_t const&, bool, signal_t*, ql::datum_t*, std::string*) at real_reql_cluster_interface.cc:484 (discriminator 1)
       10: artificial_reql_cluster_interface_t::table_reconfigure(counted_t<ql::db_t const>, name_string_t const&, table_generate_config_params_t const&, bool, signal_t*, ql::datum_t*, std::string*) at artificial_reql_cluster_interface.cc:145
       11: ql::reconfigure_term_t::eval_impl(ql::scope_env_t*, ql::args_t*, ql::eval_flags_t) const at db_table.cc:395 (discriminator 1)
       12: ql::op_term_t::term_eval(ql::scope_env_t*, ql::eval_flags_t) const at op.cc:214
       13: ql::term_t::eval(ql::scope_env_t*, ql::eval_flags_t) const at term.cc:397
       14: ql::run(ql::protob_t<Query>, rdb_context_t*, signal_t*, ql::stream_cache_t*, Response*) at term.cc:247
       15: rdb_query_server_t::run_query(ql::protob_t<Query> const&, Response*, client_context_t*) at query_server.cc:71 (discriminator 4)
       16: query_server_t::handle(http_req_t const&, http_res_t*, signal_t*) at protob.cc:447
       17: routing_http_app_t::handle(http_req_t const&, http_res_t*, signal_t*) at routing_app.cc:29 (discriminator 6)
       18: routing_http_app_t::handle(http_req_t const&, http_res_t*, signal_t*) at routing_app.cc:29 (discriminator 6)
       19: http_server_t::handle_conn(scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t) at http.cc:470
       20: boost::_mfi::mf2<void, http_server_t, scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t>::operator()(http_server_t*, scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t) const at mem_fn_template.hpp:280 (discriminator 6)
       21: void boost::_bi::list3<boost::_bi::value<http_server_t*>, boost::arg<1>, boost::_bi::value<auto_drainer_t::lock_t> >::operator()<boost::_mfi::mf2<void, http_server_t, scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t>, boost::_bi::list1<scoped_ptr_t<linux_tcp_conn_descriptor_t>&> >(boost::_bi::type<void>, boost::_mfi::mf2<void, http_server_t, scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t>&, boost::_bi::list1<scoped_ptr_t<linux_tcp_conn_descriptor_t>&>&, int) at bind.hpp:392 (discriminator 4)
       22: void boost::_bi::bind_t<void, boost::_mfi::mf2<void, http_server_t, scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t>, boost::_bi::list3<boost::_bi::value<http_server_t*>, boost::arg<1>, boost::_bi::value<auto_drainer_t::lock_t> > >::operator()<scoped_ptr_t<linux_tcp_conn_descriptor_t> >(scoped_ptr_t<linux_tcp_conn_descriptor_t>&) at bind_template.hpp:32
       23: std::_Function_handler<void (scoped_ptr_t<linux_tcp_conn_descriptor_t>&), boost::_bi::bind_t<void, boost::_mfi::mf2<void, http_server_t, scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t>, boost::_bi::list3<boost::_bi::value<http_server_t*>, boost::arg<1>, boost::_bi::value<auto_drainer_t::lock_t> > > >::_M_invoke(std::_Any_data const&, scoped_ptr_t<linux_tcp_conn_descriptor_t>&) at functional:2041
       24: std::function<void (scoped_ptr_t<linux_tcp_conn_descriptor_t>&)>::operator()(scoped_ptr_t<linux_tcp_conn_descriptor_t>&) const at functional:2440
       25: linux_nonthrowing_tcp_listener_t::handle(int) at network.cc:939 (discriminator 2)
       26: void std::_Mem_fn<void (linux_nonthrowing_tcp_listener_t::*)(int)>::operator()<int&, void>(linux_nonthrowing_tcp_listener_t*, int&) const at functional:569 (discriminator 4)
       27: void std::_Bind<std::_Mem_fn<void (linux_nonthrowing_tcp_listener_t::*)(int)> (linux_nonthrowing_tcp_listener_t*, int)>::__call<void, , 0ul, 1ul>(std::tuple<>&&, std::_Index_tuple<0ul, 1ul>) at functional:1265
       28: void std::_Bind<std::_Mem_fn<void (linux_nonthrowing_tcp_listener_t::*)(int)> (linux_nonthrowing_tcp_listener_t*, int)>::operator()<, void>() at functional:1323
       29: callable_action_instance_t<std::_Bind<std::_Mem_fn<void (linux_nonthrowing_tcp_listener_t::*)(int)> (linux_nonthrowing_tcp_listener_t*, int)> >::run_action() at callable_action.hpp:31
       30: callable_action_wrapper_t::run() at runtime_utils.cc:43
       31: coro_t::run() at coroutines.cc:199
       32: void coro_t::spawn_now_dangerously<std::_Bind<std::_Mem_fn<void (linux_nonthrowing_tcp_listener_t::*)(int)> (linux_nonthrowing_tcp_listener_t*, int)> >(std::_Bind<std::_Mem_fn<void (linux_nonthrowing_tcp_listener_t::*)(int)> (linux_nonthrowing_tcp_listener_t*, int)>&&) at coroutines.hpp:52
       33: linux_nonthrowing_tcp_listener_t::accept_loop(auto_drainer_t::lock_t) at network.cc:907
       34: void std::_Mem_fn<void (linux_nonthrowing_tcp_listener_t::*)(auto_drainer_t::lock_t)>::operator()<auto_drainer_t::lock_t&, void>(linux_nonthrowing_tcp_listener_t*, auto_drainer_t::lock_t&) const at functional:569 (discriminator 6)
       35: void std::_Bind<std::_Mem_fn<void (linux_nonthrowing_tcp_listener_t::*)(auto_drainer_t::lock_t)> (linux_nonthrowing_tcp_listener_t*, auto_drainer_t::lock_t)>::__call<void, , 0ul, 1ul>(std::tuple<>&&, std::_Index_tuple<0ul, 1ul>) at functional:1265
       36: void std::_Bind<std::_Mem_fn<void (linux_nonthrowing_tcp_listener_t::*)(auto_drainer_t::lock_t)> (linux_nonthrowing_tcp_listener_t*, auto_drainer_t::lock_t)>::operator()<, void>() at functional:1323
       37: callable_action_instance_t<std::_Bind<std::_Mem_fn<void (linux_nonthrowing_tcp_listener_t::*)(auto_drainer_t::lock_t)> (linux_nonthrowing_tcp_listener_t*, auto_drainer_t::lock_t)> >::run_action() at callable_action.hpp:31
       38: callable_action_wrapper_t::run() at runtime_utils.cc:43
       39: coro_t::run() at coroutines.cc:199
       40: coro_t* coro_t::spawn_sometime<std::_Bind<std::_Mem_fn<void (linux_nonthrowing_tcp_listener_t::*)(auto_drainer_t::lock_t)> (linux_nonthrowing_tcp_listener_t*, auto_drainer_t::lock_t)> >(std::_Bind<std::_Mem_fn<void (linux_nonthrowing_tcp_listener_t::*)(auto_drainer_t::lock_t)> (linux_nonthrowing_tcp_listener_t*, auto_drainer_t::lock_t)>&&) at coroutines.hpp:58
       41: linux_nonthrowing_tcp_listener_t::begin_listening() at network.cc:686 (discriminator 4)
       42: linux_tcp_listener_t::linux_tcp_listener_t(std::set<ip_address_t, std::less<ip_address_t>, std::allocator<ip_address_t> > const&, int, std::function<void (scoped_ptr_t<linux_tcp_conn_descriptor_t>&)> const&) at network.cc:972 (discriminator 1)
       43: http_server_t::http_server_t(std::set<ip_address_t, std::less<ip_address_t>, std::allocator<ip_address_t> > const&, int, http_app_t*) at http.cc:349 (discriminator 9)
       44: administrative_http_server_manager_t::administrative_http_server_manager_t(std::set<ip_address_t, std::less<ip_address_t>, std::allocator<ip_address_t> > const&, int, mailbox_manager_t*, boost::shared_ptr<semilattice_readwrite_view_t<cluster_semilattice_metadata_t> >, clone_ptr_t<watchable_t<change_tracking_map_t<peer_id_t, cluster_directory_metadata_t> > >, admin_tracker_t*, http_app_t*, std::string) at server.cc:264 (discriminator 4)
       45: do_serve(io_backender_t*, bool, base_path_t const&, metadata_persistence::cluster_persistent_file_t*, metadata_persistence::auth_persistent_file_t*, unsigned long, serve_info_t const&, os_signal_cond_t*) at serve.cc:368 (discriminator 7)
       46: serve(io_backender_t*, base_path_t const&, metadata_persistence::cluster_persistent_file_t*, metadata_persistence::auth_persistent_file_t*, unsigned long, serve_info_t const&, os_signal_cond_t*) at serve.cc:445
       47: run_rethinkdb_serve(base_path_t const&, serve_info_t*, file_direct_io_mode_t, int, unsigned long, uuid_u const*, cluster_semilattice_metadata_t const*, directory_lock_t*, bool*) at command_line.cc:816 (discriminator 2)
       48: run_rethinkdb_porcelain(base_path_t const&, name_string_t const&, std::set<name_string_t, std::less<name_string_t>, std::allocator<name_string_t> > const&, file_direct_io_mode_t, int, unsigned long, bool, serve_info_t*, directory_lock_t*, bool*) at command_line.cc:841
       49: void std::_Bind<void (*(base_path_t, name_string_t, std::set<name_string_t, std::less<name_string_t>, std::allocator<name_string_t> >, file_direct_io_mode_t, int, unsigned long, bool, serve_info_t*, directory_lock_t*, bool*))(base_path_t const&, name_string_t const&, std::set<name_string_t, std::less<name_string_t>, std::allocator<name_string_t> > const&, file_direct_io_mode_t, int, unsigned long, bool, serve_info_t*, directory_lock_t*, bool*)>::__call<void, , 0ul, 1ul, 2ul, 3ul, 4ul, 5ul, 6ul, 7ul, 8ul, 9ul>(std::tuple<>&&, std::_Index_tuple<0ul, 1ul, 2ul, 3ul, 4ul, 5ul, 6ul, 7ul, 8ul, 9ul>) at functional:1264
       50: void std::_Bind<void (*(base_path_t, name_string_t, std::set<name_string_t, std::less<name_string_t>, std::allocator<name_string_t> >, file_direct_io_mode_t, int, unsigned long, bool, serve_info_t*, directory_lock_t*, bool*))(base_path_t const&, name_string_t const&, std::set<name_string_t, std::less<name_string_t>, std::allocator<name_string_t> > const&, file_direct_io_mode_t, int, unsigned long, bool, serve_info_t*, directory_lock_t*, bool*)>::operator()<, void>() at functional:1323
       51: std::_Function_handler<void (), std::_Bind<void (*(base_path_t, name_string_t, std::set<name_string_t, std::less<name_string_t>, std::allocator<name_string_t> >, file_direct_io_mode_t, int, unsigned long, bool, serve_info_t*, directory_lock_t*, bool*))(base_path_t const&, name_string_t const&, std::set<name_string_t, std::less<name_string_t>, std::allocator<name_string_t> > const&, file_direct_io_mode_t, int, unsigned long, bool, serve_info_t*, directory_lock_t*, bool*)> >::_M_invoke(std::_Any_data const&) at functional:2041
       52: std::function<void ()>::operator()() const at functional:2440
       53: starter_t::run_wrapper(std::function<void ()> const&) at runtime.cc:61
       54: void std::_Mem_fn<void (starter_t::*)(std::function<void ()> const&)>::operator()<std::function<void ()>&, void>(starter_t*, std::function<void ()>&) const at functional:569 (discriminator 4)
       55: void std::_Bind<std::_Mem_fn<void (starter_t::*)(std::function<void ()> const&)> (starter_t*, std::function<void ()>)>::__call<void, , 0ul, 1ul>(std::tuple<>&&, std::_Index_tuple<0ul, 1ul>) at functional:1265
       56: void std::_Bind<std::_Mem_fn<void (starter_t::*)(std::function<void ()> const&)> (starter_t*, std::function<void ()>)>::operator()<, void>() at functional:1323
       57: std::_Function_handler<void (), std::_Bind<std::_Mem_fn<void (starter_t::*)(std::function<void ()> const&)> (starter_t*, std::function<void ()>)> >::_M_invoke(std::_Any_data const&) at functional:2041
       58: std::function<void ()>::operator()() const at functional:2440
       59: callable_action_instance_t<std::function<void ()>&>::run_action() at callable_action.hpp:31
       60: callable_action_wrapper_t::run() at runtime_utils.cc:43
       61: coro_t::run() at coroutines.cc:199
       62: coro_t* coro_t::spawn_sometime<std::function<void ()>&>(std::function<void ()>&) at coroutines.hpp:58
       63: starter_t::on_thread_switch() at runtime.cc:57
       64: linux_message_hub_t::on_event(int) at message_hub.cc:154
       65: epoll_event_queue_t::run() at epoll.cc:101 (discriminator 2)
       66: linux_thread_pool_t::start_thread(void*) at thread_pool.cc:160
       67: /usr/lib/libpthread.so.0(+0x7124) [0x7f630b09b124] at 0x7f630b09b124 (/usr/lib/libpthread.so.0)
       68: clone+0x6d at 0x7f630adcf4bd (/usr/lib/libc.so.6)
error: Exiting.
```

**Comment by Tryneus** [comment]
Just reproduced this with the shard fuzzer, so it still exists in the codebase.  Here is an updated backtrace:

```c
Version: rethinkdb 1.15.2-1743-gb2034c (debug) (GCC 4.6.3)
error: Error in ../src/clustering/administration/tables/split_points.cc at line 72:
error: Guarantee failed: [split_points->at(i) >= split_points->at(i-1)]
error: Backtrace:
error: Fri Jan 16 13:44:56 2015

       1: rethinkdb_backtrace(void**, int) at rethinkdb_backtrace.cc:101
       2: backtrace_t::backtrace_t() at backtrace.cc:203
       3: lazy_backtrace_formatter_t::lazy_backtrace_formatter_t() at backtrace.cc:283
       4: format_backtrace(bool) at backtrace.cc:198
       5: report_fatal_error(char const*, int, char const*, ...) at errors.cc:83
       6: /home/ssd2/grey/rethinkdb/build/debug_system/rethinkdb() [0x1b75e94] at 0x1b75e94 ()
       7: calculate_split_points_with_distribution(std::map<store_key_t, long, std::less<store_key_t>, std::allocator<std::pair<store_key_t const, long> > > const&, unsigned long, table_shard_scheme_t*, std::string*) at split_points.cc:157
       8: calculate_split_points_intelligently(uuid_u, real_reql_cluster_interface_t*, unsigned long, table_shard_scheme_t const&, signal_t*, table_shard_scheme_t*, std::string*) at split_points.cc:234
       9: real_reql_cluster_interface_t::reconfigure_internal(cluster_semilattice_metadata_t*, counted_t<ql::db_t const> const&, uuid_u const&, name_string_t const&, table_generate_config_params_t const&, bool, signal_t*, ql::datum_t*, std::string*) at real_reql_cluster_interface.cc:709
       10: real_reql_cluster_interface_t::table_reconfigure(counted_t<ql::db_t const>, name_string_t const&, table_generate_config_params_t const&, bool, signal_t*, ql::datum_t*, std::string*) at real_reql_cluster_interface.cc:790
       11: artificial_reql_cluster_interface_t::table_reconfigure(counted_t<ql::db_t const>, name_string_t const&, table_generate_config_params_t const&, bool, signal_t*, ql::datum_t*, std::string*) at artificial_reql_cluster_interface.cc:235
       12: ql::reconfigure_term_t::eval_impl_on_table_or_db(ql::scope_env_t*, ql::args_t*, ql::eval_flags_t, counted_t<ql::db_t const> const&, boost::optional<name_string_t> const&) const at db_table.cc:493
       13: ql::table_or_db_meta_term_t::eval_impl(ql::scope_env_t*, ql::args_t*, ql::eval_flags_t) const at db_table.cc:366
       14: ql::op_term_t::term_eval(ql::scope_env_t*, ql::eval_flags_t) const at op.cc:217
       15: ql::runtime_term_t::eval(ql::scope_env_t*, ql::eval_flags_t) const at term.cc:438
       16: ql::run(ql::protob_t<Query>, rdb_context_t*, signal_t*, ql::stream_cache_t*, ip_and_port_t const&, Response*) at term.cc:271
       17: rdb_query_server_t::run_query(ql::protob_t<Query> const&, Response*, client_context_t*, ip_and_port_t const&) at query_server.cc:61
       18: void query_server_t::connection_loop<json_protocol_t>(linux_tcp_conn_t*, client_context_t*) at protob.cc:315
       19: query_server_t::handle_conn(scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t) at protob.cc:277
       20: std::_Mem_fn<void (query_server_t::*)(scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t)>::operator()(query_server_t*, scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t) const at functional:551
       21: void std::_Bind<std::_Mem_fn<void (query_server_t::*)(scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t)> (query_server_t*, std::_Placeholder<1>, auto_drainer_t::lock_t)>::__call<void, scoped_ptr_t<linux_tcp_conn_descriptor_t>&, 0, 1, 2>(std::tuple<scoped_ptr_t<linux_tcp_conn_descriptor_t>&>&&, std::_Index_tuple<0, 1, 2>) at functional:1146
       22: void std::_Bind<std::_Mem_fn<void (query_server_t::*)(scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t)> (query_server_t*, std::_Placeholder<1>, auto_drainer_t::lock_t)>::operator()<scoped_ptr_t<linux_tcp_conn_descriptor_t>&, void>(scoped_ptr_t<linux_tcp_conn_descriptor_t>&&&) at functional:1206
       23: std::_Function_handler<void (scoped_ptr_t<linux_tcp_conn_descriptor_t>&), std::_Bind<std::_Mem_fn<void (query_server_t::*)(scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t)> (query_server_t*, std::_Placeholder<1>, auto_drainer_t::lock_t)> >::_M_invoke(std::_Any_data const&, scoped_ptr_t<linux_tcp_conn_descriptor_t>&) at functional:1780
       24: std::function<void (scoped_ptr_t<linux_tcp_conn_descriptor_t>&)>::operator()(scoped_ptr_t<linux_tcp_conn_descriptor_t>&) const at functional:2162
       25: linux_nonthrowing_tcp_listener_t::handle(int) at network.cc:930
```

**Comment by danielmewes** [comment]
@timmaxw can you take another look at this now that @Tryneus has a way to reproduce it?

**Comment by Tryneus** [comment]
@danielmewes - ah, saying that I can reproduce it is a stretch.  I'm working on the shard fuzzer still, but it is not yet to the point that I can give a deterministic history of the table's configuration.

**Comment by Tryneus** [comment]
Update: I have reproduced this again, and from the core file I've made a unittest that generates this error consistently.  I'll work on simplifying the test parameters, then investigate the sharding algorithm.

**Comment by Tryneus** [comment]
Looks like everything is working fine except that `interpolate_key` is broken.  Here is the test case (I tried to simplify it more, but all changes I made caused the bug to go away):

Distribution:
```
distribution[""] = 1
distribution["N"] = 1
distribution["NZabc"] = 6
distribution["NZabc\x00"] = 6
distribution["Nc"] = 6
distribution["S1"] = 6
distribution["S2"] = 6
distribution["S3"] = 6
distribution["S4"] = 6
```

I see the following from the rebalancing:
```
interpolating with left pair: "NZabc", right pair: "NZabc\x00", ratio: 0.3333
interpolated key before clamp: "NZabc\x00\x00\x00\x00"
resulting key: "NZabc\x00"

interpolating with left pair: "NZabc\x00", right pair: "Nc", ratio: 0.0000
interpolated key before clamp: "NcZab"
resulting key: "Nc"

interpolating with left pair: "NZabc\x00", right pair: "Nc", ratio: 0.8333
interpolated key before clamp: "N\x10a\x90\x3B"
resulting key: "NZabc\x00"

interpolating with left pair: "Nc", right pair: "S1", ratio: 0.5000
interpolated key before clamp: "\x00P\xCA\x00"
resulting key: "Nc"

interpolating with left pair: "S1", right pair: "S2", ratio: 0.1667
interpolated key before clamp: "S\xAA1\x2A\xAA"
resulting key: "S2"

interpolating with left pair: "S2", right pair: "S3", ratio: 0.0000
interpolated key before clamp: "S\x002\x00\x00"
resulting key: "S2"

interpolating with left pair: "S2", right pair: "S3", ratio: 0.6667
interpolated key before clamp: "S\xAA2\xAA\xAA"
resulting key: "S3"

interpolating with left pair: "S3", right pair: "S4", ratio: 0.3333
interpolated key before clamp: "SU3UU"
resulting key: "S4"
```

As you can see, none of the calls to `interpolate_key` resulted in a key that fell in-between the left and right bounds, so the resulting key was then clamped to one of the bounds.  In some cases, this can cause the algorithm to flip-flop between the bounds of a single range and cause a non-monotonically-increasing sharding scheme.

Haven't investigated the `interpolate_key` algorithm yet, will get to that tomorrow.

**Comment by timmaxw** [comment]
The bug in `interpolate_key()` might be in this loop:
```c++
/* Append the interpolated value onto `out_buf` */
for (size_t k = 0; k < num_interp; ++k) {
    uint8_t c = (out_tail >> (8 * (num_interp - k))) & 0xFF;
    if (i + k < MAX_KEY_SIZE) {
        out_buf[i + k] = c;
    }
}
```
I tihnk it should be `8 * (num_interp - k - 1)` instead of `8 * (num_interp - k)`. Currently, it will shift `out_tail` by 32, then 24, then 16, then 8 bits. It should be shifting by 24, then 16, then 8, then 0 bits. So the first byte of the interpolated block will be derived from shifting a 32-bit value by 32 bits, which is an undefined operation. I think this is consistent with the pattern of errors that you posted.

**Comment by Tryneus** [comment]
Since there was no issue for it, I'll put it here: the `tests/interface/shard_fuzzer.py` test has been added and merged to `next` in commit 867b83a, and cherry-picked into `v1.16.x` in commit 58bf8a9.

**Comment by Tryneus** [comment]
The fix for `interpolate_key()` is up in review 2506.  @timmaxw, your analysis was correct, and that fixed the crash.  I also added behavior to strip trailing null-bytes from the key (before clamping), to avoid situations like:

```
key_rebalance('a', 'b', 0.0) => 'a\x00\x00\x00'
```

**Comment by Tryneus** [comment]
The fix has been approved and merged to `next` in commits bed14ea and 92cf909, and cherry-picked into `v1.16.x` in commits fbe2462 and 0e3578c.  Will be in release 1.16.0.
